include/silo/common/bidirectional_map.h

```diff
@@ -30,14 +29,45 @@ class BidirectionalMap {
    std::unordered_map<V, Idx> value_to_id;
 
   public:
-   BidirectionalMap();
-   BidirectionalMap(BidirectionalMap&& map) = delete;
+   BidirectionalMap()
+       : id_to_value(),
+         value_to_id() {}
 
-   [[nodiscard]] V getValue(Idx idx) const;
+   BidirectionalMap(BidirectionalMap&& map) = default;
+   BidirectionalMap& operator=(BidirectionalMap&& map) = default;
 
-   [[maybe_unused]] [[nodiscard]] std::optional<Idx> getId(V value) const;
+   // Copy constructor and operator are deleted to restrict mistaken copies of this data-structure
+   // When this data-structure is copied, the two copies' dictionary will diverge over-time
+   // If you want to copy this data-structure, explicitly call BidirectionalMap::copy() and
+   // make sure that the contained type's copy is permissible for your design.
+   BidirectionalMap(const BidirectionalMap& map) = delete;
+   BidirectionalMap& operator=(const BidirectionalMap& map) = delete;
 
-   [[nodiscard]] Idx getOrCreateId(V value);
+   BidirectionalMap copy() const {
+      BidirectionalMap result;
+      result.id_to_value = id_to_value;
+      result.value_to_id = value_to_id;
+      return result;
+   }
+
+   V getValue(Idx idx) const { return id_to_value.at(idx); }
+
+   std::optional<Idx> getId(const V& value) const {
+      if (value_to_id.contains(value)) {
+         return value_to_id.at(value);
+      }
+      return std::nullopt;
+   }
+
+   Idx getOrCreateId(V value) {
+      if (value_to_id.contains(value)) {
+         return value_to_id.at(value);
+      }
+      const Idx identifier = id_to_value.size();
+      id_to_value.push_back(value);
+      value_to_id[value] = identifier;
+      return identifier;
+   }
```

Comment:
As this is templated code, it will not lead to duplicate symbols and has the additional advantage that the compiler will be helped at inlining :)
This follows the [google style guide](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers):
> When a header declares inline functions or templates that clients of the header will instantiate, the inline functions and templates must also have definitions in the header, either directly or in files it includes.

Note that definition=implementation